feature/experimental/p4rt/ate_tests/performance_test/performance_test.go

```diff
@@ -0,0 +1,932 @@
+	ate          *ondatra.ATEDevice
+	top          *ondatra.ATETopology
+	gdpPacketIO  PacketIO
+	lldpPacketIO PacketIO
+	trPacketIO   PacketIO
+}
+
+type PacketIOPacket struct {
+	TTL            *uint8
+	SrcMAC, DstMAC *string
+	EthernetType   *uint32
+	HopLimit       *uint8
+}
+
+type PacketIO interface {
+	GetPacketTemplate() *PacketIOPacket
+	GetEgressPort() string
+	GetIngressPort() string
+}
+
+type LLDPPacketIO struct {
+	PacketIOPacket
+	IngressPort string
+}
+
+type GDPPacketIO struct {
+	PacketIOPacket
+	IngressPort string
+}
+
+type TraceroutePacketIO struct {
+	PacketIOPacket
+	IngressPort string
+	EgressPort  string
+}
+
+// programTableEntry programs or deletes p4rt table entry based on delete flag.
+func programTableEntry(client *p4rt_client.P4RTClient, delete bool) error {
+	updateType := p4_v1.Update_INSERT
+	if delete {
+		updateType = p4_v1.Update_DELETE
+	}
+	return client.Write(&p4_v1.WriteRequest{
+		DeviceId:   deviceID,
+		ElectionId: &p4_v1.Uint128{High: uint64(0), Low: electionID},
+		Updates: p4rtutils.ACLWbbIngressTableEntryGet(
+			getTableEntry(updateType),
+		),
+		Atomicity: p4_v1.WriteRequest_CONTINUE_ON_ERROR,
+	})
+}
+
+// testTraffic sends ATE traffic, stop and collect total packet tx from ATE source port.
+func testTraffic(t *testing.T, ate *ondatra.ATEDevice, flows []*ondatra.Flow, srcEndPoint *ondatra.Interface, duration int) int {
+	t.Helper()
+	for _, flow := range flows {
+		flow.WithSrcEndpoints(srcEndPoint).WithDstEndpoints(srcEndPoint)
+	}
+	ate.Traffic().Start(t, flows...)
+	time.Sleep(time.Duration(duration) * time.Second)
+	ate.Traffic().Stop(t)
+
+	outPkts := gnmi.GetAll(t, ate, gnmi.OC().FlowAny().Counters().OutPkts().State())
+	total := 0
+	for _, count := range outPkts {
+		total += int(count)
+	}
+	return total
+}
+
+// sendPackets GDP/LLDP/Traceroute sends out packets via PacketOut message in StreamChannel.
+func sendPackets(t *testing.T, client *p4rt_client.P4RTClient, packet *p4_v1.PacketOut, packetCount int, delay float64) {
+	for i := 0; i < packetCount; i++ {
+		if err := client.StreamChannelSendMsg(
+			&streamName, &p4_v1.StreamMessageRequest{
+				Update: &p4_v1.StreamMessageRequest_Packet{
+					Packet: packet,
+				},
+			}); err != nil {
+			t.Errorf("There is error seen in Packet Out. %v, %s", err, err)
+		}
+		time.Sleep(time.Duration(delay) * time.Millisecond)
+	}
+}
+
+// decodeIPPacketTTL decodes L2 header in the packet and returns TTL, packetData[14:0] to remove first 14 bytes of Ethernet header.
+func decodeIPPacketTTL(packetData []byte) uint8 {
+	packet := gopacket.NewPacket(packetData, layers.LayerTypeEthernet, gopacket.Default)
+	etherHeader := packet.Layer(layers.LayerTypeEthernet)
+	if etherHeader != nil {
+		header := etherHeader.(*layers.Ethernet)
+		if header.EthernetType == layers.EthernetType(0x0800) {
+			packet1 := gopacket.NewPacket(packetData[14:], layers.LayerTypeIPv4, gopacket.Default)
+			IPv4 := packet1.Layer(layers.LayerTypeIPv4)
```

Comment:
local variables should be lowercase, and you might as well group this into the if-statement,i.e.
```
if ipv4 := packet1.Layer(layers.LayerTypeIPv4); ipvr != nil {
	return ipv4.(*layers.IPv4).TTL
}
```